- **Kelas** : IF403
- **Program Studi** : PJJ Informatika
- **Nama Mahasiswa** : Raka Anggie Saputra
- **NIM** : 240401010148

In [1]:
import pandas as pd

df = pd.read_csv("telco_churn.csv")
print(df.shape)
print(df["Churn"].value_counts(normalize=True))

(7043, 21)
Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64


In [4]:
from sklearn.model_selection import train_test_split

# Encoding fitur kategorikal
df = pd.get_dummies(df, drop_first=True)

# Pisahkan fitur (X) dan target (y)
X = df.drop("Churn_Yes", axis=1)
y = df["Churn_Yes"]

X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

In [5]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(
    n_estimators=300,
    class_weight="balanced",
    random_state=42
)

rf.fit(X_tr, y_tr)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

In [6]:
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score

# Hitung prediksi
y_pred = rf.predict(X_te)

# Hitung probabilitas kelas churn
y_prob = rf.predict_proba(X_te)[:, 1]

# Tampilkan hasil evaluasi
print(classification_report(y_te, y_pred))

print("ROC-AUC:", roc_auc_score(y_te, y_prob))

              precision    recall  f1-score   support

       False       0.84      0.90      0.87      1035
        True       0.65      0.52      0.58       374

    accuracy                           0.80      1409
   macro avg       0.74      0.71      0.72      1409
weighted avg       0.79      0.80      0.79      1409

ROC-AUC: 0.8337260068717869


In [8]:
# Hitung probabilitas churn
prob_churn = rf.predict_proba(X_te)[:, 1]

# Tampilkan 10 probabilitas pertama
print(prob_churn[:10])

[0.00333333 0.63666667 0.07333333 0.37666667 0.01666667 0.56666667
 0.46333333 0.03       0.00333333 0.53      ]


Model Random Forest menunjukkan performa yang baik dalam memprediksi pelanggan yang berpotensi melakukan churn berdasarkan nilai precision, recall, F1-score, dan ROC-AUC yang diperoleh. Selain menghasilkan prediksi kelas, model juga memberikan probabilitas churn melalui `predict_proba`, sehingga tingkat risiko setiap pelanggan dapat diketahui. Pelanggan dengan probabilitas mendekati 1 memiliki risiko churn yang lebih tinggi dan dapat diprioritaskan untuk program retensi. Dengan demikian, model dapat membantu perusahaan dalam mengambil keputusan yang lebih tepat untuk mengurangi tingkat churn.


## Kesimpulan

Pada Pertemuan 10 dilakukan **prediksi customer churn** menggunakan algoritma **Random Forest** pada dataset Telco Customer Churn. Dataset memiliki **7.043 data dan 21 atribut**, dengan sekitar **26,5% pelanggan mengalami churn** dan 73,5% tidak mengalami churn. 

Model Random Forest dibuat menggunakan **300 decision trees** dan `class_weight="balanced"` untuk membantu menangani ketidakseimbangan kelas.

Hasil evaluasi model:

* **Accuracy:** 80%
* **Precision churn:** 65%
* **Recall churn:** 52%
* **F1-Score churn:** 58%
* **ROC-AUC:** 0,834 

Model juga dapat menghasilkan **probabilitas churn** menggunakan `predict_proba`. Semakin mendekati 1 nilai probabilitasnya, semakin tinggi risiko pelanggan tersebut untuk melakukan churn. 

Random Forest mampu melakukan prediksi churn dengan performa yang cukup baik dan tidak hanya memberikan hasil klasifikasi, tetapi juga **tingkat risiko churn setiap pelanggan**. Informasi tersebut dapat digunakan perusahaan untuk menentukan pelanggan yang perlu diprioritaskan dalam program retensi dan mengurangi tingkat customer churn. 
